### PR TITLE
Add unit test for Calender.BusinessDaysBetween method.

### LINKED ISSUE #21: Add tests to Dates.Calender
Context
It is best practices to have 100% code coverage.

Details
Add tests for QuantSA.Shared.Dates.Calender

### PR DESCRIPTION
**Pull request by rwx-oberholzer-marcelle**
# ✅ Pull Request Checklist

Before submitting your pull request, please ensure the following are complete:

## 🔍 Self Review
- [x] Confirmed that the destination branch is correct.
- [x] I have reviewed my own code carefully.
- [x] I have removed any debugging code or print statements.
- [x] Code changes are scoped to the purpose of this PR (no unrelated changes).
- [x] Comments are clear, and the code is self-explanatory where possible.
- [x] Documentation (if applicable) is updated.

## 📄 Task
- [ ] The task/issue title is clear and descriptive.
- [ ] The description explains **why** (context) and **what** (details) clearly.
- [ ] Related issues are referenced (e.g., "Fixes #123").

## ✅ Code Quality
- [ ] There are no new warnings or errors in the build output.
- [ ] Code follows the project’s coding standards and formatting rules.

## 🧪 Testing
- [ ] All unit and integration tests pass locally.
- [ ] New functionality is covered by tests, or is explicitly marked as not testable.
- [ ] Edge cases and error handling have been considered.

## ✅ CI
- [ ] CI checks are passing (e.g., build, tests, linting).

## 📄 PR Quality
- [ ] The PR title is clear and descriptive.
- [ ] The PR description explains **what** and **why** clearly.
- [ ] Related issues are referenced (e.g., "Fixes #123").

## 👥 Collaboration
- [x] I have assigned reviewers and/or requested feedback.
- [x] I’ve addressed or responded to all prior review comments (if applicable).

### CODE REVIEW

File: QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    [TestClass]
+    public class CalenderTests
+    {
+        [TestMethod]
+        public void BusinessDaysBetween_ExcludesWeekends()
+        {
+            var calender = new Calendar("TestCalender");
+            var start = new Date(2024, 6, 3);
+            var end = new Date(2024, 6, 7);
+            int result = calender.BusinessDaysBetween(start, end);
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs]: Use var instead of int.

File: QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+namespace QuantSA.Shared.Test.Dates
+{
+    [TestClass]
+    public class CalenderTests
+    {
+        [TestMethod]
+        public void BusinessDaysBetween_ExcludesWeekends()
+        {
+            var calender = new Calendar("TestCalender");
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs]: Spelling "calendar".

File: QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
+﻿using System.Globalization;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using QuantSA.Shared.Dates;
+using Calendar = QuantSA.Shared.Dates.Calendar;
+
+namespace QuantSA.Shared.Test.Dates
+{
+    [TestClass]
+    public class CalenderTests
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs]: Spelling "Calendar"

File: QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    {
+        [TestMethod]
+        public void BusinessDaysBetween_ExcludesWeekends()
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs]: Find a better name for the test.

File: QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs (hunk 1 above)
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs]: Add line breaks between: Initialisation of inputs, call to get result, and assertions.

File: QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs (hunk 2 above)
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs]: If needed for other tests, move this into the setup method.

**Review by rwx-jacquier-robert — CHANGES_REQUESTED** [review]
Get 100% coverage, if possible.

File: QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs (hunk 2 above)
Comment (MarcelleOberholzer) [reply QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs:14]: Upon further investigation, the correct spelling is indeed calendar.

Calendar and calender are two words that are often confused due to their similar spelling. However, a calendar is a system for organizing days, typically for social, religious, commercial, or administrative purposes, arranged by days, weeks, and months within a year. In contrast, a calender is a machine in which paper or cloth is smoothed and compressed by passing between rollers. It's essential to use these terms correctly to avoid confusion and to convey the intended meaning.

File: QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+        private Date _end;
+
+        [TestInitialize]
+        public void Setup()
+        {
+            _calendar = new Calendar("TestCalender");
+            _start = new Date(2024, 6, 17);
+            _end = new Date(2024, 6, 21);
+
+            //this is needed for the IsHoliday test
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs]: No need for this comment.

Also, when you do comment, the style is "// This is a comment". Note the space after the //

File: QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+            _calendar.AddHoliday(_start);
+
+            //this is needed for the EndOfMonth test
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs]: Not needed.

File: QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+            var holiday = new Date(2024, 6, 28);
+            _calendar.AddHoliday(holiday);
+        }
+
+        [TestMethod]
+        public void IsHoliday_ReturnsCorrect()
+        {
+            Assert.IsFalse(_calendar.IsHoliday(_end));
+            Assert.IsTrue(_calendar.IsHoliday(_start));
+        }
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs]: Split into two tests and name them appropriately. The thinking here is that if something breaks, you want to have a fine-grained idea of what broke. When you start throwing multiple test cases into a test you lose some of this granularity.

File: QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs:1]: I'm not a fan of the "ReturnsCorrectly" naming, since its a bit ambiguous. Rather use "ReturnsExpectedResult"

File: QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+
+        [TestMethod]
+        public void IsBusinessDay_ReturnsCorrectly()
+        {
+            var weekday = new Date(2024, 6, 18);
+            var weekend = new Date(2024, 6, 22);
+
+            Assert.IsFalse(_calendar.IsBusinessDay(_start));
+            Assert.IsTrue(_calendar.IsBusinessDay(weekday));
+            Assert.IsFalse(_calendar.IsBusinessDay(weekend));
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs]: Split into multiple tests.

File: QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+        [TestMethod]
+        public void IsWeekend_ReturnsCorrectly()
+        {
+            var saturday = DayOfWeek.Saturday;
+            var sunday = DayOfWeek.Sunday;
+            var monday = DayOfWeek.Monday;
+
+            Assert.IsTrue(Calendar.IsWeekend(saturday));
+            Assert.IsTrue(Calendar.IsWeekend(sunday));
+            Assert.IsFalse(Calendar.IsWeekend(monday));
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs]: Split into multiple tests.

File: QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+
+        [TestMethod]
+        public void BusinessDaysBetween_ReturnsCorrectly()
+        {
+            var from = new Date(2024, 6, 17); 
+            var to = new Date(2024, 6, 21);   
+
+            var result = _calendar.BusinessDaysBetween(from, to);
+
+            Assert.AreEqual(3, result); //this is because the BusinessDaysBetween method by default includes the from and excludes the to.
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs]: Comment style. Use sentence case and leave a space after the //

**Review by rwx-jacquier-robert — CHANGES_REQUESTED** [review]

File: QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+            var weekday = new Date(2024, 6, 18);
+
+            Assert.IsTrue(_calendar.IsBusinessDay(weekday));
+        }
+
+        [TestMethod]
+        public void IsBusinessDay_ReturnsWeekendAsExpected()
+        {
+            var weekend = new Date(2024, 6, 22);
+
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs]: Unnecessary newline.

File: QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+            _end = new Date(2024, 6, 21);
+
+            _calendar.AddHoliday(_start);
+
+            var holiday = new Date(2024, 6, 28);
+            _calendar.AddHoliday(holiday);
+        }
+
+        [TestMethod]
+        public void IsHoliday_ReturnsTrueCorrect()
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs]: Test naming is still an issue, for instance, this one should be "IsHoliday_DayIsHoliday_ReturnsTrue".

That way you're specifying the state (DayIsHoliday), and the "correctly" part at the end, which is redundant, is removed.

**Review by rwx-jacquier-robert — CHANGES_REQUESTED** [review]
Look at the comments I added, revisit all the test names and refactor them.

File: QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs (hunk 3 above)
Comment (rwx-jacquier-robert) [reply QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs]: @MarcelleOberholzer Still not resolved. Please don't mark the conversations as resolved, I'll do that.

File: QuantSA/QuantSA.Shared/Dates/Calendar.cs
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Shared/Dates/Calendar.cs:1]: Spelling of file name.

File: QuantSA/QuantSA.Shared/Dates/Calendar.cs [hunk 13]
```diff
@@ -80,15 +80,22 @@ public bool IsHoliday(Date date)
         /// </summary>
         public bool IsEndOfMonth(Date d)
         {
-            throw new NotImplementedException();
+            return d == EndOfMonth(d);
         }
 
         /// <summary>
         /// last business day of the month to which the given date belongs
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Shared/Dates/Calendar.cs]: Capitalisation of Last

File: QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+
+            var calendar = new Calendar("HolidayCalendar", holidays);
+
+            Assert.IsTrue(calendar.IsHoliday(new Date(2024, 12, 25)));
+            Assert.IsTrue(calendar.IsHoliday(new Date(2024, 1, 1)));
+            Assert.AreEqual("HolidayCalendar", calendar.GetName());
+        }
+
+        [TestMethod]
+        public void GetName_CalendarIsTestMarket_ReturnsTestMarket()
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs]: GetName_ReturnsExpectedValue is fine for this test name.

File: QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+            var from = new Date(2024, 6, 21);
+            var to = new Date(2024, 6, 17);  
+
+            var result = _calendar.BusinessDaysBetween(from, to); 
+
+            Assert.AreEqual(-4, result); 
+        }
+
+        [TestMethod]
+        public void Constructor_WithHolidays_AddsToCalendar()
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs]: "Constructor_WithHolidays_ReturnsExpected" value is fine.

File: QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+            var from = new Date(2024, 6, 17); 
+            var to = new Date(2024, 6, 21);   
+
+            var result = _calendar.BusinessDaysBetween(from, to);
+
+            Assert.AreEqual(3, result); // This is because the BusinessDaysBetween method by default includes the from and excludes the to.
+        }
+
+        [TestMethod]
+        public void BusinessDaysBetween_21To17Jun2024_ReturnsNegative4()
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs]: Too specific, doesn't add value. "BusinessDaysBetween_FromMoreThanTo_ReturnsExpectedNegativeValue" is fine.

File: QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+        [TestMethod]
+        public void IsWeekend_Sunday_ReturnsFalse()
+        {
+            var sunday = DayOfWeek.Sunday;
+
+            Assert.IsTrue(Calendar.IsWeekend(sunday));
+        }
+
+        [TestMethod]
+        public void BusinessDaysBetween_17To21Jun2024_Returns3()
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs]: Too specific, doesn't add value. "BusinessDaysBetween_FromLessThanTo_ReturnsExpectedValue" is fine.

File: QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+        public void IsEndOfMonth_31Jan2024_ReturnsTrue()
+        {
+            var isLastDay = new Date(2024, 1, 31);
+            var actualEndOfMonth = _calendar.EndOfMonth(isLastDay);
+
+            Assert.IsTrue(_calendar.IsEndOfMonth(isLastDay));
+        }
+
+        [TestMethod]
+        public void EndOfMonth_15Jun2024_Returns27Jun2024()
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs]: Too specific, see other comments to get a better name.

File: QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs [hunk 19]
```diff
@@ -0,0 +1,150 @@
...
+        }
+
+        [TestMethod]
+        public void IsHoliday_DaysIsNotHoliday_ReturnsFalse()
+        {
+            Assert.IsFalse(_calendar.IsHoliday(_end));
+        }
+
+        [TestMethod]
+        public void IsEndOfMonth_31Jan2024_ReturnsTrue()
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Shared.Test/Dates/CalenderTests.cs]: Too specific, see other comments to get a better name.

**Review by rwx-jacquier-robert — CHANGES_REQUESTED** [review]

File: QuantSA/QuantSA.Shared.Test/Dates/CalendarTests.cs [hunk 20]
```diff
@@ -0,0 +1,150 @@
...
+            var from = new Date(2024, 6, 21);
+            var to = new Date(2024, 6, 17);  
+
+            var result = _calendar.BusinessDaysBetween(from, to); 
+
+            Assert.AreEqual(-4, result); 
+        }
+
+        [TestMethod]
+        public void Constructor_WithHolidays_ReturnsExpected()
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Shared.Test/Dates/CalendarTests.cs]: ReturnsExpectedValue is better.

File: QuantSA/QuantSA.Shared.Test/QuantSA.Shared.Test.csproj [hunk 21]
```diff
@@ -9,8 +9,8 @@
 
   <ItemGroup>
     <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.14.1" />
-    <PackageReference Include="MSTest.TestAdapter" Version="2.1.2" />
-    <PackageReference Include="MSTest.TestFramework" Version="2.1.2" />
+    <PackageReference Include="MSTest.TestAdapter" Version="3.9.3" />
+    <PackageReference Include="MSTest.TestFramework" Version="3.9.3" />
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Shared.Test/QuantSA.Shared.Test.csproj:13]: Where are these changes coming from?

**Review by rwx-jacquier-robert — CHANGES_REQUESTED** [review]

File: QuantSA/QuantSA.Shared.Test/QuantSA.Shared.Test.csproj (hunk 21 above)
Comment (rwx-oberholzer-marcelle) [reply QuantSA/QuantSA.Shared.Test/QuantSA.Shared.Test.csproj:13]: @robertjacquier-rw 

It's the same as with the tests added to Dates. I had issues running the tests and updated the versions on NuGet.
